File: coordinator/benchmark.py
"""
Proof-of-Hardware benchmark using PyTorch matrix multiplication.
This script is called by worker nodes to prove their hardware capabilities.
"""

import hashlib
import sys
import time

import torch


def get_device():
    """Detect the best available hardware."""
    if torch.cuda.is_available():
        return torch.device("cuda"), torch.cuda.get_device_name(0)
    elif torch.backends.mps.is_available():
        return torch.device("mps"), "Apple Metal (MPS)"
    else:
        return torch.device("cpu"), "CPU (Slow)"


def run_proof_of_hardware(challenge_seed: str, matrix_size=4096, iterations=100):
    """
    Run the benchmark challenge.

    Args:
        challenge_seed: Random seed from coordinator
        matrix_size: Size of matrices (default 4096x4096)
        iterations: Number of iterations (default 100)

    Returns:
        tuple: (proof_hash, duration_seconds, device_name)
    """
    device, device_name = get_device()

    # Deterministic seeding
    seed_int = int(hashlib.sha256(challenge_seed.encode()).hexdigest(), 16) % (2**32)
    torch.manual_seed(seed_int)

    # Initialize matrices
    try:
        a = torch.randn(matrix_size, matrix_size, device=device)
        b = torch.randn(matrix_size, matrix_size, device=device)
    except RuntimeError as e:
        return None, 0, f"OOM Error: {str(e)}"

    # Synchronize before timing
    if device.type == "cuda":
        torch.cuda.synchronize()

    start_time = time.time()

    # The benchmark loop
    current_val = torch.tensor([0.0], device=device)

    for i in range(iterations):
        # Heavy matrix multiplication
        c = torch.matmul(a, b)

        # Chain dependency to prevent optimization
        slice_val = c[0, 0]
        a.add_(slice_val * 0.00001)
        current_val.add_(slice_val)

    # Synchronize after computation
    if device.type == "cuda":
        torch.cuda.synchronize()

    duration = time.time() - start_time

    # Generate proof hash
    final_val_str = str(current_val.item())
    proof_hash = hashlib.sha256(final_val_str.encode()).hexdigest()

    return proof_hash, duration, device_name


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python benchmark.py <challenge_seed>")
        sys.exit(1)

    challenge_seed = sys.argv[1]

    print("--- 🚀 STARTING BENCHMARK ---", file=sys.stderr)
    print(f"Seed: {challenge_seed}", file=sys.stderr)

    proof, duration, device = run_proof_of_hardware(challenge_seed)

    if proof:
        print("--- ✅ RESULTS ---", file=sys.stderr)
        print(f"Device: {device}", file=sys.stderr)
        print(f"Time: {duration:.4f}s", file=sys.stderr)
        print(f"Proof: {proof}", file=sys.stderr)

        # Output JSON to stdout for parsing
        import json

        result = {"proof_hash": proof, "duration": duration, "device_name": device}
        print(json.dumps(result))
    else:
        print("❌ Benchmark failed", file=sys.stderr)
        sys.exit(1)

File: worker/src/infrastructure/system/benchmark.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::time::Duration;
use tokio::process::Command;
use tracing::{error, info, warn};

#[derive(Debug, Serialize, Deserialize)]
pub struct BenchmarkResult {
    pub proof_hash: String,
    pub duration: f64,
    pub device_name: String,
}

#[derive(Debug, Deserialize)]
struct BenchmarkOutput {
    proof_hash: String,
    duration: f64,
    device: String,
}

/// Run hardware benchmark using Python subprocess
pub async fn run_benchmark(seed: &str, matrix_size: usize) -> Result<BenchmarkResult> {
    info!(
        "🔬 Starting hardware benchmark (seed: {}, size: {})",
        seed, matrix_size
    );

    // Spawn Python subprocess
    // The benchmark.py is at the root of the worker directory
    let output = tokio::time::timeout(
        Duration::from_secs(300), // 5 minute timeout
        Command::new("python3")
            .arg("benchmark.py")
            .arg(seed)
            .arg(matrix_size.to_string())
            .output(),
    )
    .await
    .context("Benchmark timed out after 300 seconds")?
    .context("Failed to execute benchmark subprocess")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        error!("Benchmark failed: {}", stderr);

        // Check if it's a PyTorch import error
        if stderr.contains("No module named 'torch'") {
            warn!("PyTorch not installed, falling back to CPU benchmark");
            return run_cpu_fallback_benchmark(seed, matrix_size).await;
        }

        anyhow::bail!("Benchmark subprocess failed: {stderr}");
    }

    // Parse JSON output
    let stdout = String::from_utf8_lossy(&output.stdout);
    let benchmark_output: BenchmarkOutput =
        serde_json::from_str(&stdout).context("Failed to parse benchmark JSON output")?;

    info!(
        "✓ Benchmark complete: {}s on {}",
        benchmark_output.duration, benchmark_output.device
    );

    Ok(BenchmarkResult {
        proof_hash: benchmark_output.proof_hash,
        duration: benchmark_output.duration,
        device_name: benchmark_output.device,
    })
}

/// Fallback CPU benchmark when GPU/PyTorch unavailable
async fn run_cpu_fallback_benchmark(seed: &str, matrix_size: usize) -> Result<BenchmarkResult> {
    info!("Running CPU fallback benchmark...");

    // Simple CPU benchmark using numpy
    let python_code = r#"
import json
import hashlib
import time
import sys
import numpy as np

seed = sys.argv[1]
matrix_size = int(sys.argv[2])

# Set seed for reproducibility
np.random.seed(int(seed, 16) % (2**32))

# Generate matrices
start = time.time()
a = np.random.randn(matrix_size, matrix_size).astype(np.float32)
b = np.random.randn(matrix_size, matrix_size).astype(np.float32)

# Perform matrix multiplication
result = np.matmul(a, b)

duration = time.time() - start

# Generate proof hash
proof_data = f"{seed}:{duration:.6f}:{result.sum():.6f}"
proof_hash = hashlib.sha256(proof_data.encode()).hexdigest()

output = {
    "proof_hash": proof_hash,
    "duration": duration,
    "device": "CPU (fallback)"
}

print(json.dumps(output))
"#;

    let output = tokio::time::timeout(
        Duration::from_secs(300),
        Command::new("python3")
            .arg("-c")
            .arg(python_code)
            .arg(seed)
            .arg(matrix_size.to_string())
            .output(),
    )
    .await
    .context("CPU fallback benchmark timed out")?
    .context("Failed to execute CPU fallback")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("CPU fallback failed: {stderr}");
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let benchmark_output: BenchmarkOutput =
        serde_json::from_str(&stdout).context("Failed to parse CPU fallback output")?;

    Ok(BenchmarkResult {
        proof_hash: benchmark_output.proof_hash,
        duration: benchmark_output.duration,
        device_name: benchmark_output.device,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_run_cpu_fallback_benchmark_success() {
        // This test requires python3 and numpy to be available in the environment
        let result = run_cpu_fallback_benchmark("test-seed", 128).await;
        if let Ok(res) = result {
            assert!(!res.proof_hash.is_empty());
            assert!(res.duration > 0.0);
            assert_eq!(res.device_name, "CPU (fallback)");
        }
    }

    #[tokio::test]
    async fn test_run_benchmark_not_found() {
        // Test that it handles missing benchmark.py
        let result = run_benchmark("test-seed", 128).await;
        assert!(result.is_err());
    }
}
